src/compiler/crystal/codegen/link.cr

```diff
@@ -105,83 +110,68 @@ module Crystal
-
-            if static && (static_lib = find_static_lib(libname, CrystalLibraryPath.paths))
-              flags << ' ' << static_lib
-            elsif has_pkg_config && (libflags = pkg_config_flags(libname, static, library_path))
-              flags << ' ' << libflags
-            elsif static && (static_lib = find_static_lib(libname, library_path))
-              flags << ' ' << static_lib
-            else
-              flags << " -l" << libname
-            end
-          end
+      # Add CRYSTAL_LIBRARY_PATH locations, so the linker preferentially
+      # searches user-given library paths.
+      CrystalLibraryPath.paths.each do |path|
+        flags << "'-L#{path}'"
+      end
 
-          if framework = ann.framework
-            flags << " -framework " << framework
-          end
+      link_annotations.reverse_each do |ann|
+        if ldflags = ann.ldflags
+          flags << ldflags
         end
 
-        # Append the CRYSTAL_LIBRARY_PATH values as -L flags.
-        CrystalLibraryPath.paths.each do |path|
-          flags << " -L#{path}"
+        # First, check pkg-config for the pkg-config module name if provided, then
+        # check pkg-config with the lib name, then fall back to -lname
+        if (pkg_config_name = ann.pkg_config) && (flag = pkg_config(pkg_config_name, static_build))
+          flags << flag
+        elsif (lib_name = ann.lib) && (flag = pkg_config(lib_name, static_build))
+          flags << flag
+        elsif (lib_name = ann.lib)
+          flags << "-l#{lib_name}"
         end
-        # Append the default paths as -L flags in case the linker doesn't know
-        # about them (eg: FreeBSD won't search /usr/local/lib by default):
-        library_path.each do |path|
-          flags << " -L#{path}"
+
+        if framework = ann.framework
+          flags << "-framework" << framework
         end
       end
-    end
 
-    def link_annotations
-      annotations = [] of LinkAnnotation
-      add_link_annotations @types, annotations
-      annotations
+      flags.join(" ")
     end
 
-    private def pkg_config_flags(libname, static, library_path)
-      if system("pkg-config #{libname}")
-        if static
-          flags = [] of String
-          `pkg-config #{libname} --libs --static`.split.each do |cfg|
-            if cfg.starts_with?("-L")
-              library_path << cfg[2..-1]
-            elsif cfg.starts_with?("-l")
-              flags << (find_static_lib(cfg[2..-1], library_path) || cfg)
-            else
-              flags << cfg
-            end
-          end
-          flags.join ' '
-        else
-          `pkg-config #{libname} --libs`.chomp
-        end
+    PKG_CONFIG_PATH = Process.find_executable("pkg-config")
+
+    # Returns the result of running `pkg-config mod` but returns nil if
+    # pkg-config is not installed, or the module does not exist.
+    private def pkg_config(mod, static = false) : String?
+      return unless pkg_config_path = PKG_CONFIG_PATH
```

Comment:
Stylistic: It is really hard to follow the control flow when the actual effect (i.e. assignment to local var) happens in a trailing conditional.
```suggestion
      pkg_config_path = PKG_CONFIG_PATH || return
```
`return unless pkg_config_path` could also be on a separate line.

Reply:
thanks i hate it